### PR TITLE
Driver: led: lp5562: Added en_gpio DTS property and Implemented PM Support

### PR DESCRIPTION
**Pull request by cx-anuj-pathak**
### Description
This PR attempts to add LP5562 EN Pin handling if connected to MCU's GPIO. 

### Changes
- Added optional en_gpios DT property to lp5562.yaml
- Added lp5562_enable/disable API to handle EN pin toggle and EN bit set in right order with appropriate delays
- Added PM_DEVICE support using lp5562_enable/disable APIs.

### CODE REVIEW

File: drivers/led/lp5562.c
Comment (ubieda) [inline drivers/led/lp5562.c:1098]: For PM Device to work, we need to add the PM Device instance, then include it in the Device DT Instance. See: https://github.com/zephyrproject-rtos/zephyr/blob/b2475804f8502d21c9d7aeddfd35daa764892fdd/drivers/sensor/lm77/lm77.c#L414-L421

**Review by ubieda — CHANGES_REQUESTED** [review]
NACKs:
- Let's complete the PM Device support.
- Let's split the PM Device and the EN_GPIO support on separate patches.

File: drivers/led/lp5562.c
Comment (ubieda) [inline drivers/led/lp5562.c:929]:
```suggestion
	static int lp5562_enable(const struct device *dev, bool soft_reset)
```

**Review by ubieda — DISMISSED** [review]

**Review by ubieda — DISMISSED** [review]

File: drivers/led/lp5562.c
Comment (simonguinot) [inline drivers/led/lp5562.c:966]: Well, it is written "500 µs" in the comment :)

File: dts/bindings/led/ti,lp5562.yaml
Comment (simonguinot) [inline dts/bindings/led/ti,lp5562.yaml:29]: Is that a dual-purpose pin as on the LP5569 chip ? If not then I wonder if it would not be better to add a fixed regulator in the devicetree configuration of your board. See https://github.com/zephyrproject-rtos/zephyr/pull/76713#pullrequestreview-2233320574 for reference.

Also I think you should rename `en-gpios` into `enable-gpios`. And this addition should be mentioned into the release notes.

And finally, I think you can squash this commit with the next one.

File: drivers/led/lp5562.c
Comment (bbilas) [inline drivers/led/lp5562.c:1075]: not needed

File: drivers/led/lp5562.c
Comment (bbilas) [inline drivers/led/lp5562.c:1010]: You should propagate errno from function, `ENODEV` is not suitable here.

File: drivers/led/lp5562.c
Comment (simonguinot) [inline drivers/led/lp5562.c]: Are there any changes above related to the ? If it's just a reindentation, please put it in a dedicated commit.

File: drivers/led/lp5562.c
Comment (simonguinot) [inline drivers/led/lp5562.c:935]: Please use a capital letter to start sentence and a period to end it. We want to be consistent with other comments in the. Also please use GPIO instead of gpio (accronym).

My comment above applies to other parts.

**Review by simonguinot — COMMENTED** [review]
Hi @cx-anuj-pathak,

Thanks for this patch. See my comments below.

File: dts/bindings/led/ti,lp5562.yaml
Comment (RickBruyninckx) [reply dts/bindings/led/ti,lp5562.yaml:29]: The datasheet says "VCC/EN" so it only has one purpose. The fixed regulator is actually how I solved it for our project. (It gave us the flexibility of enabling/disabling multiple devices with one output).

But in this implementation the enable-gpios are optional so a user can choose, right?

File: drivers/led/lp5562.c
Comment (RickBruyninckx) [inline drivers/led/lp5562.c:974]: Shouldn't there be a NULL check? or do we assume that's handled by writing to the device?

File: drivers/led/lp5562.c
Comment (RickBruyninckx) [inline drivers/led/lp5562.c:985]: what if you want to disable the led driver but continue communication (for writing a new colorloop for example)?

File: drivers/led/lp5562.c
Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c:966]: `k_usleep` "is unlikely to work as expected without kernel tuning" as per api documentation. thus used 1ms.
update comment to explain in detail

File: dts/bindings/led/ti,lp5562.yaml
Comment (cx-anuj-pathak) [reply dts/bindings/led/ti,lp5562.yaml:29]: @simonguinot updated the name, but i don't know where to add it in release notes, Could you point me to the location?

and yes `enable-gpios` is single purpose for 5562, and is optional.

File: drivers/led/lp5562.c
Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c]: Compliance CI warn about formatting of diff to be compatible with clang-format. That's why formatted the diff.
I will omit it.

File: drivers/led/lp5562.c
Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c:974]: I didn't get it, which NULL check is missing could you explain?

```
	const struct lp5562_config *config = dev->config;
	const struct gpio_dt_spec *en_pin = &config->en_pin;
```
these lines are used everywhere without check and `en_pin` handling is inside NULL check.

File: drivers/led/lp5562.c
Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c:985]: Do you mean SUSPEND/RESUME should only handle LP5562_ENABLE EN Bit, while TURN_ON/OFF should toggle EN_GPIO ?
If so, IMO EN_GPIO is a hardware suspend method not the turn on/off. and that is why TURN_ON/OFF is not supported.

Let me know if my understanding of review comment is incorrect or you do not agree

**Review by RickBruyninckx — APPROVED** [review]
The code looks good

**Review by ubieda — DISMISSED** [review]

File: dts/bindings/led/ti,lp5562.yaml
Comment (ubieda) [reply dts/bindings/led/ti,lp5562.yaml]: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-4.0.rst

File: drivers/led/lp5562.c
Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c:990]: resolving it since got approval

File: dts/bindings/led/ti,lp5562.yaml
Comment (cx-anuj-pathak) [reply dts/bindings/led/ti,lp5562.yaml]: I can see it after rebasing, and updated it.
Thanks!

File: dts/bindings/led/ti,lp5562.yaml
Comment (bbilas) [inline dts/bindings/led/ti,lp5562.yaml:1]: I would rename the commit title to `dts: bindings: lp5562: Added en-gpio property`

File: drivers/led/lp5562.c
Comment (simonguinot) [reply drivers/led/lp5562.c:983]: There are two cases here:

1. the system clock rate is high enough and `k_usleep` will work as expected
2. the system clock rate is too low and `k_usleep` will wake up at the next tick boundary

Both cases are OK. So let's use `k_usleep`.

Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c]: changed to k_sleep(K_USEC(500));

**Review by bbilas — DISMISSED** [review]

**Review by ubieda — DISMISSED** [review]

File: dts/bindings/led/ti,lp5562.yaml
Comment (simonguinot) [reply dts/bindings/led/ti,lp5562.yaml]: > The datasheet says "VCC/EN" so it only has one purpose. The fixed regulator is actually how I solved it for our project. (It gave us the flexibility of enabling/disabling multiple devices with one output).
> 
> But in this implementation the enable-gpios are optional so a user can choose, right?

Well, it looks weird to me to duplicate the same code in every driver to handle power through a GPIO.

File: drivers/led/lp5562.c
Comment (simonguinot) [inline drivers/led/lp5562.c:1035]: I think you can get rid of this log message. Indeed all the error cases are already logged in `lp5562_enable()` . So this generic message brings nothing more.

**Review by simonguinot — CHANGES_REQUESTED** [review]

File: drivers/led/lp5562.c
Comment (bbilas) [reply drivers/led/lp5562.c:987]: It looks like we need to extend the `tests/drivers/build_all/led` with PM enabled.

Comment (simonguinot) [reply drivers/led/lp5562.c:987]: Indeed :)

Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c:987]: @simonguinot should I add a commit with @bbilas suggestion in this PR?

Comment (simonguinot) [reply drivers/led/lp5562.c:987]: No thanks. This should go through another PR. I'll take care of it.

Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c]: pushed changed and this time tested both with/without CONFIG_PM_DEVICE, compilation

File: drivers/led/lp5562.c
Comment (simonguinot) [reply drivers/led/lp5562.c]: OK it is a detail but I doesn't like "EN GPIO". This doesn't match anything. In the LP5562 datasheet, the pin is called EN/VCC. And in the DT binding, the GPIO is called "enable-gpios".  So IMHO we should try to refer to one of them in the log.

So maybe:

LOG_ERR("%s: failed to set EN/VCC GPIO to 0", dev->name);

Or:

LOG_ERR("%s: failed to set enable GPIO to 0", dev->name);

Comment (simonguinot) [reply drivers/led/lp5562.c]: And sorry for missing that the GPIO is set to 0 with my initial proposition !

Comment (cx-anuj-pathak) [reply drivers/led/lp5562.c]: Okay used dts reference instead

**Review by bbilas — APPROVED** [review]

**Review by thedjnK — APPROVED** [review]

**Review by simonguinot — COMMENTED** [review]
Thanks for this PR @cx-anuj-pathak,

After discussing with other maintainers it appears that using an `enable-gpios` DT property is OK. EN/VCC is a dedicated pin of the LP5562 controller. There is no regulator here. So using a fixed regulator to handle this pin would be convenient but would not describe the hardware correctly. The `enable-gpios` property does a better job in that regard.

So let's not wait any longer to get this PR merged !

Please, can you fix your commit messages ? I can read en-gpio, en_gpio and EN-GPIO. Please replace them with the actual property name (enable-gpio) or the hardware pin name (EN/VCC) if appropriate.

Thanks !

**Review by simonguinot — APPROVED** [review]
Thanks @cx-anuj-pathak !

**Pull request merged**
